### PR TITLE
Allow import/export patterns as JSON files

### LINKED ISSUE #53288: Patterns: Allow import/export in Site Editor to match wp-admin
As we are leaning into the Site Editor's Patterns page when available, it makes sense to provide similar functionality to the wp-admin page for patterns (formerly reusable blocks). One missing piece at the moment is the ability to import or export patterns from the Site Editor's Patterns page.

### PR DESCRIPTION
**Pull request by kevin940726**
## What?

Close #53288. Allow import/export patterns as JSON files in the site editor's pattern screen.

## Why?

To match wp-admin functionality.

## How?

The implementation is quite straightforward. Just add some menu items to the corresponding dropdown menus.

## Testing Instructions

1. Create some synced/unsynced patterns.
2. Visit Site Editor -> Patterns -> My patterns. Tab through the grid list to the "Actions" menu dropdown of a pattern, and select "Export as JSON".
3. You should be prompted by a native download dialog to decide where to download the JSON file.
4. Verify that the downloaded JSON file is the same as the one downloaded in `/wp-admin/edit.php?post_type=wp_block`.
5. Go back to Site Editor -> Patterns
6. Press <kbd>Tab</kbd> once to navigate to the "Create pattern" button. Hit <kbd>Enter</kbd> to open the dropdown, and select "Import pattern from JSON"
7. You should be prompted by a native upload dialog that only accepts a `.json` file.
8. Choose a validated pattern JSON file to import it to the patterns screen. There should also be a snackbar announcing the successful message `Imported "title" from JSON.`.
9. Choosing an invalid pattern JSON file should announce "Invalid JSON file" in the snackbar.

### Testing Instructions for Keyboard

Same as above.

## Screenshots or screencast 

https://github.com/WordPress/gutenberg/assets/7753001/22e965c8-07c8-4fba-8e91-0a4fdc91e293

### CODE REVIEW

File: packages/edit-site/src/components/page-patterns/grid-item.js [hunk 1]
```diff
@@ -252,6 +268,12 @@ function GridItem( { categoryId, item, ...props } ) {
 										: __( 'Duplicate' )
 								}
 							/>
+							{ item.type === USER_PATTERNS && (
+								<MenuItem onClick={ () => exportAsJSON() }>
```
Comment (kevin940726) [inline packages/edit-site/src/components/page-patterns/grid-item.js:296]: Question: Should we close the dropdown menu after it's pressed? I feel like that might be doing too much. I could use some accessibility feedback here 🤔 .

File: packages/patterns/src/store/actions.js [hunk 2]
```diff
@@ -40,18 +34,43 @@ export const __experimentalCreatePattern =
...
-		const newBlock = createBlock( 'core/block', {
-			ref: updatedRecord.id,
-		} );
-		registry
-			.dispatch( blockEditorStore )
-			.replaceBlocks( clientIds, newBlock );
-		dispatch.__experimentalSetEditingPattern( newBlock.clientId, true );
-		return updatedRecord;
+		const pattern = await dispatch.__experimentalCreatePattern(
+			parsedContent.title,
```
Comment (kevin940726) [inline packages/patterns/src/store/actions.js:76]: Question: Should we increment the title if there's a same title already in the patterns list?

File: packages/edit-site/src/components/add-new-pattern/index.js [hunk 3]
```diff
@@ -29,6 +33,11 @@ export default function AddNewPattern() {
 		const settings = select( editSiteStore ).getSettings();
 		return !! settings.supportsTemplatePartsMode;
 	}, [] );
+	const { __experimentalCreatePatternFromFile: createPatternFromFile } =
```
Comment (aaronrobertshaw) [inline packages/edit-site/src/components/add-new-pattern/index.js]: I think we should probably avoid exporting the `__experimental` prefixed API and instead export it [via private-apis](https://github.com/WordPress/gutenberg/blob/f9e405e0e53d61cd36af4f3b34f2de75874de1e1/docs/contributors/code/coding-guidelines.md#private-selectors-and-actions). What do you think?

File: packages/edit-site/src/components/page-patterns/grid-item.js (hunk 1 above)
Comment (aaronrobertshaw) [reply packages/edit-site/src/components/page-patterns/grid-item.js:296]: I'm no a11y expert so I'll defer to others here. One point to note though is that the same dropdown is closed when we rename a pattern etc so it should probably be consistent one way or the other.

File: packages/patterns/src/store/actions.js (hunk 2 above)
Comment (aaronrobertshaw) [reply packages/patterns/src/store/actions.js:76]: Ultimately, having unique pattern names would be a good thing although to date there hasn't been anything enforcing it. There's already an issue open for this in https://github.com/WordPress/gutenberg/issues/52360. 

Unique names would help when we come to introducing slugs so they are consistent across all types of patterns.

**Review by aaronrobertshaw — COMMENTED** [review]
Thanks for working on this @kevin940726 👍 

Tests nicely for me!

✅ Verified that exported patterns match between wp-admin and site editor 
✅ Importing patterns works well

The only issue I spotted was a minor implementation detail. We should probably leverage private apis to export `createPatternFromFile` from the patterns package.

The `__experimental` prefix from all the other actions etc are being removed in https://github.com/WordPress/gutenberg/pull/54338.

File: packages/edit-site/src/components/add-new-pattern/index.js (hunk 3 above)
Comment (kevin940726) [reply packages/edit-site/src/components/add-new-pattern/index.js]: Oh I wasn't aware of that PR! Just rebased and updated it [5ffa599](https://github.com/WordPress/gutenberg/pull/54337/commits/5ffa599e3e014a882b370a742140dde5ac17aeda)!

File: packages/patterns/src/store/actions.js (hunk 2 above)
Comment (kevin940726) [reply packages/patterns/src/store/actions.js:76]: I'm thinking if we ever want to enforce the uniqueness then it should be done in the backend. I guess it's better to leave it for future PRs. WDYT?

**Comment by kevin940726** [comment]
@SaxonF Just to confirm do you think this is a good first approach for adding import/export features to patterns?

File: packages/patterns/src/store/actions.js (hunk 2 above)
Comment (aaronrobertshaw) [reply packages/patterns/src/store/actions.js:76]: Agreed. We'll focus on that in a follow and can discuss options via the dedicated issue.

File: packages/edit-site/src/components/page-patterns/grid-item.js [hunk 4]
```diff
@@ -108,6 +110,20 @@ function GridItem( { categoryId, item, ...props } ) {
 	};
 	const deleteItem = () =>
 		isTemplatePart ? removeTemplate( item ) : deletePattern();
+	const exportAsJSON = () => {
+		const json = {
+			__file: item.type,
+			title: item.title,
+			content: item.reusableBlock.content.raw,
+			syncStatus: item.reusableBlock.wp_pattern_sync_status,
```
Comment (aaronrobertshaw) [inline packages/edit-site/src/components/page-patterns/grid-item.js]:
```suggestion
			content: item.patternBlock.content.raw,
			syncStatus: item.patternBlock.wp_pattern_sync_status,
```

After the rebase the `reusableBlock` property doesn't exist so this throws an error. The suggested tweak fixes it for me.

File: packages/patterns/src/store/actions.js [hunk 5]
```diff
@@ -47,18 +41,43 @@ export const createPattern =
...
 
-		const newBlock = createBlock( 'core/block', {
-			ref: updatedRecord.id,
-		} );
-		registry
-			.dispatch( blockEditorStore )
-			.replaceBlocks( clientIds, newBlock );
-		dispatch.setEditingPattern( newBlock.clientId, true );
-		return updatedRecord;
+		const pattern = await dispatch.__experimentalCreatePattern(
```
Comment (aaronrobertshaw) [inline packages/patterns/src/store/actions.js]:
```suggestion
		const pattern = await dispatch.createPattern(
```

This needs updating after the rebase as well.

Screenshot 2023-09-20 at 3 12 58 pm

File: packages/edit-site/src/components/add-new-pattern/index.js [hunk 6]
```diff
@@ -64,6 +72,13 @@ export default function AddNewPattern() {
 			onClick: () => setShowPatternModal( true ),
 			title: __( 'Create pattern' ),
 		},
+		{
+			icon: symbol,
+			onClick: () => {
+				patternUploadInputRef.current.click();
+			},
+			title: __( 'Import pattern from JSON' ),
```
Comment (aaronrobertshaw) [inline packages/edit-site/src/components/add-new-pattern/index.js]: This option probably needs to be last in the dropdown menu.

**Review by aaronrobertshaw — COMMENTED** [review]
Thanks for updating this PR @kevin940726 👍 

There are a couple of issues after the rebase that break the intended functionality.

While testing I also wondered if we should be exporting the pattern's categories too. 

I was on a category other than "All patterns" or "Uncategorized" (the one belonging to the pattern I just tested exporting). Importing after that doesn't give much indication of where your imported pattern went, just the snackbar notice that it was imported.

The category export/import could be left to a follow-up, if it is in fact desired. What do you think?

I also ran into another edge case. If you try to import the same pattern file multiple times nothing happens. If you then delete the imported pattern and try again the import still fails without error. Refreshing the page and trying again the pattern import works.

We should also move the Import menu option to the bottom as it's the least likely to be what the user wants.

Screenshot 2023-09-20 at 3 26 29 pm

Given we are short on time before the beta for 6.4 is cut, this could be pushed to 6.5 if needed.

**Comment by kevin940726** [comment]
> There are a couple of issues after the rebase that break the intended functionality.

Thanks for spotting them!

> I was on a category other than "All patterns" or "Uncategorized" (the one belonging to the pattern I just tested exporting). Importing after that doesn't give much indication of where your imported pattern went, just the snackbar notice that it was imported.

This is fixed by always navigating to the `all-patterns` category after importing. We could instead navigate the users to the specific category when we support importing/exporting categories too!

> The category export/import could be left to a follow-up, if it is in fact desired. What do you think?

Yeah, I'm not sure if we want that in the aspect of product or UX. We can discuss it in a follow-up if needed 👍 .

> I also ran into another edge case. If you try to import the same pattern file multiple times nothing happens. If you then delete the imported pattern and try again the import still fails without error. Refreshing the page and trying again the pattern import works.

Ahh right. This is a bug and it's now fixed 👍 .

> We should also move the Import menu option to the bottom as it's the least likely to be what the user wants.

Moved to the bottom 👍 .

> Given we are short on time before the beta for 6.4 is cut, this could be pushed to 6.5 if needed.

Agreed 👍

**Comment by kevin940726** [comment]
> This is fixed by always navigating to the `all-patterns` category after importing. We could instead navigate the users to the specific category when we support importing/exporting categories too!

Hmm, on another thought, I think we could instead auto-apply the current category when importing. That might be what the users expect anyway 🤔.

**Comment by aaronrobertshaw** [comment]
> Hmm, on another thought, I think we could instead auto-apply the current category when importing. That might be what the users expect anyway 🤔.

Do you think we have enough time to make this change and still land this for 6.4? Could applying the current category be a follow-up?

**Comment by kevin940726** [comment]
> Do you think we have enough time to make this change and still land this for 6.4? Could applying the current category be a follow-up?

I think it's only a few lines change and I can do it by today.

**Comment by kevin940726** [comment]
> Hmm, on another thought, I think we could instead auto-apply the current category when importing. That might be what the users expect anyway 🤔.

Done in [9a625d8](https://github.com/WordPress/gutenberg/pull/54337/commits/9a625d8476d5b3396b339a516cb2881294862bf5).

**Review by aaronrobertshaw — APPROVED** [review]
Thanks for the iterations @kevin940726 👍 

> Done in [9a625d8](https://github.com/WordPress/gutenberg/pull/54337/commits/9a625d8476d5b3396b339a516cb2881294862bf5).

This works well for existing user-created pattern categories but doesn't work if the user has a theme pattern category selected e.g. Featured etc.

Everything else is testing well for me. I think we should merge this and iterate on the theme pattern category issue in a follow-up, what do you think?



https://github.com/WordPress/gutenberg/assets/60436221/46d8a104-06ff-42ec-848c-8decee0d6842

**Pull request merged**